Esp8266 async uart

unaiur opened the pull request:
New try to contribute asynchronous UART code. This time, I created new methods instead of replacing the old ones. I also refactorized the synchronous version a bit to reuse as much code as possible.

If the refactorization isn´t desired, I can generate a new patchset without the refactoring.

Makuna commented:
Give me a few days to review this, I just got to glance at it and I like what I have seen so far.

corani commented on src/internal/NeoEsp8266AsyncUartMethod.h:
```diff
@@ -0,0 +1,89 @@
+
+    ~NeoEsp8266AsyncUartMethodBase()
+    {
+        while (!this->IsReadyToUpdate())
```
Is this necessary in the destructor? Seems to me this doesn't _"do"_ anything besides waiting.

unaiur replied:
It just waits until some microseconds until all the asyncPixels data is
written to the UART. This way we output the complete scene before freeing
the allocated data.
On Jun 5, 2016 08:09, "Daniel Bos" <notifications@github.com> wrote:

> In src/internal/NeoEsp8266AsyncUartMethod.h
> <https://github.com/Makuna/NeoPixelBus/pull/108#discussion_r65811818>:
>
> > +    void ICACHE_RAM_ATTR esp8266_uart1_async_prepare();
> > +    void ICACHE_RAM_ATTR esp8266_uart1_async_send_pixels(uint8_t* pixels, uint8_t* end);
> > +}
> > +
> > +template<typename T_SPEED> class NeoEsp8266AsyncUartMethodBase: public NeoEsp8266UartMethodBase<T_SPEED>
> > +{
> > +public:
> > +    NeoEsp8266AsyncUartMethodBase(uint8_t pin, uint16_t pixelCount, size_t elementSize)
> > +	: NeoEsp8266UartMethodBase<T_SPEED>(pin, pixelCount, elementSize)
> > +    {
> > +        _asyncPixels = (uint8_t*)malloc(this->_sizePixels);
> > +    }
> > +
> > +    ~NeoEsp8266AsyncUartMethodBase()
> > +    {
> > +        while (!this->IsReadyToUpdate())
>
> Is this necessary in the destructor? Seems to me this doesn't *"do"*
> anything besides waiting.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Makuna/NeoPixelBus/pull/108/files/1564354eb8f5376e98b74a7849ed257d497d187c#r65811818>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFnLB3Pyyyh2kiUt2lgVQhJUUETJPqPeks5qImgigaJpZM4IuMAL>
> .
>

Makuna commented on src/internal/NeoEsp8266UartMethod.h:
```diff
@@ -84,7 +83,7 @@ template<typename T_SPEED> class NeoEsp8266UartMethodBase
         SET_PERI_REG_MASK(UART_CONF0(UART1), (BIT(22)));
 
-        _endTime = micros();
+        _startTime = micros() - (getPixelTime() - 50); // After initialization, wait 50 microseconds to force a reset
```
What is the reason for waiting here?

unaiur replied:
Standard UART mode sets TX line to 3.3V while idle. Then we invert the UART logic levels in the line above, to have 0V while idle. This 3.3V to 0V transition can create a bit in the ws2812 bus. We wait 50 microsecond so the bus is reset and that bit is ignored.

Makuna replied:
That would be a great comment to add at that point then.

Makuna commented:
1) Use my coding style 
a) No Tabs, indent by four spaces
b) when declaring a pointer type, the asterisk goes with the type not the variable. 
``` 
char* variable; // good
char *variable; // not good
char * variable; // not good
```
c) all branches must have {} even if only one statement; and the { is always on its own line
```
if (name.length() > 7)
{
    return false;
}
```
d) use the standard integral type names
```
uint32_t variable; // good
int index; // not good
u32 color; // not good
```
e) no dashes in filenames.  Generally avoid all symbols. 
2) Please collapse all your changes into one before submitting a pull.  You can do this creating a new branch that you will create the pull from, then from the command line squash all your changes from the branch you were working in.
```
git merge --squash (working branch name)
```

unaiur commented:
I´ve refactored the UART classes to remove standalone functions. I created a non-templated base class, because I was having problems setting a template member as a interrupt handler. This forced me to add _pixelTime property, which contains the total time in microseconds needed to dump all pixels to the strip.

You can review the result in this testing branch: https://github.com/unaiur/NeoPixelBus/tree/lab2

@Makuna, If you like it, I will overwrite the esp8266-async-uart branch so you can merge it.

Makuna commented:
I created this gitter channel for us to discuss (https://gitter.im/Makuna/NeoPixelBusPull108).
I like some of the stuff, but there are some better ways to solve some issues here.  

Why is Update in the base class?  You have one in the async versions already, so why is there a need to have one there?  

InitializeUart has two semantically different meanings.  One is to initialize uart at a low level, the other is to setup interrupt handler (which also calls the other).  I prefer to keep separate the concepts of shared low level code and high level concept of identity (override of InitializeUart by each class).

Another code style thing, all branch statements (if, when, for, switch, etc) should have a space after them so its obvious they are branches  and not function calls.
```
if (true) // good
if(true) // bad
```

I don't think making the callback a member is worth adding a storage for something that is basically a constant.  I would rather leave it as a nonmember but still in the cpp file.  Thus you can retain the template.

Why are you including exceptions attribute like throw()?

unaiur commented:
1. Laziness... to avoid written a new synchronous class with just the Update() method. The base class is used for synchronous UART and the child class for asynchronous UART. We can create an extra class to have a nicer class inheritance diagram.
2. Calling both method InitializeUart allows to share NeoEsp8266UartMethod template. Base class  does common initialization, child class does additional tasks to prepare it for asynchronous operation: it not only install an interrupt handler, it also configures the Uart to generate interrupts at the correct time.
3. My personal style is very different than yours. I'm trying hardly to format the code the way you like, but I fail at times. I'll review it again.
4. I didn't realize that Arduino doesn't use exceptions. When using exceptions, marking functions that doesn't generate them with throw() generates faster & smaller code. But in Arduino it will only confuse people. I'm removing them.

unaiur commented:
I've reworked a bit the class hierarchy. Now I think we have an strong base.

The top class it's NeoEsp8266Uart which contains the low level functionality to handle the UART (semi-)synchronously. NeoEsp8266AsyncUart derives from NeoEsp8266Uart and modifies its behaviour to run the UART fully asynchronously using interrupts. It reuses much of the NeoEsp8266Uart implementation (indeed, only UpdateUart() is specific for each class).

Then we have NeoEsp8266UartMethodBase. I've added an extra parameter to this template, so it can inherit from either NeoEsp8266Uart or NeoEsp8266AsyncUart. NeoEsp8266UartMethodBase contains the timing constants and all methods needed to convert it in a valid NeoPixelBus method. It calls InitializeUart() & UpdateUart() in its base class to handle the low level parts.

I call NeoEsp8266Uart semi-synchronous because it sends asynchronously the last 128 bytes (the size of the UART fifo).

Makuna commented on src/internal/NeoEsp8266UartMethod.cpp:
```diff
@@ -0,0 +1,209 @@
+    while (pixels < end)
+    {
+        uint8_t subpix = *pixels++;
+        U1F = _uartData[(subpix >> 6) & 3];
```
I dislike the syntax sugar that this represents.  When you glance, you think why is the same variable  (U1F) being set over and over.  Its not quickly obvious that this really is pushing data into queue.  I believe there are other assessors that are more "function" like that make this quickly obvious what is happening.

unaiur replied:
Sorry, but I can't find any wrapper:
```
[unaiur@satnai framework-arduinoespressif]$ git grep U1F
cores/esp8266/esp8266_peri.h:#define U1F   ESP8266_REG(0xF00) //UART FIFO
doc/rgisters_dump.txt:0x60000F00 | 00000000000000000000000000000000 | 0x00000000 | 0          |      | U1F
[unaiur@satnai framework-arduinoespressif]$ 
```

unaiur replied:
uart_write_char() does additional functions, like checking that the TX port is enabled and the FIFO isn't full. Writting directly to the FIFO hardware register is much faster.

unaiur replied:
Added an static inline function to enqueue bytes in the TX FIFO.

Makuna commented on src/internal/NeoEsp8266UartMethod.cpp:
```diff
@@ -0,0 +1,209 @@
+    return pixels;
+}
+
+NeoEsp8266AsyncUart::NeoEsp8266AsyncUart(uint8_t pin, uint16_t pixelCount, size_t elementSize)
```
NeoEsp8266AsncUart should still live in its own file.  Is there a reason to collapse them into one file?

unaiur replied:
Only because NeoEsp8266Uart & NeoEsp8266AsynUart are closely related and they both share NeoEsp8266UartMethodBase.

unaiur replied:
@Makuna I see two logic options:
  a. Keep as is.
  b. Move NeoEsp8266Uart & NeoEsp8266AsynUart to its own file and left NeoEsp8266UartMethod.h for NeoEsp8266UartMethodBase template. This will generate 5 files:
      - NeoEsp8266Uart.h/cpp
      - NeoEsp8266AsyncUart.h/cpp
      - NeoEsp8266UartMethod.h

Makuna replied:
Leave as is for now, worse case before I make the next release I could modify it.

Makuna commented on src/internal/NeoEsp8266UartMethod.cpp:
```diff
@@ -0,0 +1,209 @@
+void NeoEsp8266AsyncUart::UpdateUart()
+{
+    // Instruct ESP8266 hardware uart1 to send the pixels asynchronously
+    esp8266_uart1_async_buf = this->_pixels;
```
this is problematic, the pixels buffer is exposed so that the consumer can manipulate it directly (it has a fixed format) for any advanced pixel manipulation.  But you are swapping the pointer on them.

unaiur replied:
The advantage in my implementation is that the copy is done concurrently, while the hardware is already writing out the bytes.

unaiur replied:
I realize that a program can initialize a pointer to the internal memory and write to it after pushing the update, but I can't find any example doing it. We can document that limitation in the description of the async methods.

Makuna commented on src/internal/NeoEsp8266UartMethod.cpp:
```diff
@@ -0,0 +1,209 @@
+
+    // Copy the pixels to the idle buffer and swap them
+    memcpy(_asyncPixels, this->_pixels, this->_sizePixels);
+    std::swap(_asyncPixels, this->_pixels);
```
Why is this swap needed?  See above comment.

unaiur replied:
The purpose is to maximize the frame rate. memcpy'ing several KB of data takes time. With this implementation, memcpy is done concurrently while UART hardware is already sending the previous frame.

Makuna commented on src/internal/NeoEsp8266UartMethod.h:
```diff
@@ -95,43 +119,34 @@ template<typename T_SPEED> class NeoEsp8266UartMethodBase
     uint32_t getPixelTime() const
     {
-        return (T_SPEED::ByteSendTimeUs * _sizePixels);
+        return (T_SPEED::ByteSendTimeUs * this->_sizePixels);
```
(not a problem, just letting you know)
generally I avoid the this-> as all member variables are prefixed with an underscore.  But I also don't require that it be omitted.

unaiur replied:
Not in this case. _sizePixel is defined in a base class that is a template parameter.

unaiur replied:
I removed all unneeded this pointers in non-templated classes

Makuna commented on src/internal/NeoEsp8266UartMethod.cpp:
```diff
@@ -0,0 +1,209 @@
+        uint8_t subpix = *pixels++;
+        U1F = _uartData[(subpix >> 6) & 3];
+        U1F = _uartData[(subpix >> 4) & 3];
+        U1F = _uartData[(subpix >> 2) & 3];
```
minor:  bit wise operations with constants should have the constant as either hex or binary; so 3 should 0x03.  I know you just copied, so not a big deal.

unaiur replied:
Replaced.
